dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,346 @@
+        return gotVal === UNDEFINED ? void 0 : JSON.parse(gotVal);
+    }
+
+    setItem(key, value) {
+        const setVal = value === void 0 ? UNDEFINED : JSON.stringify(value);
+        this._storage.setItem(storePrefix + key, setVal);
+    }
+
+    removeItem(key) {
+        this._storage.removeItem(storePrefix + key);
+    }
+
+    clear(keyPrefix) {
+        const fullPrefix = storePrefix + keyPrefix;
+        const keyMatch = keyPrefixMatch(fullPrefix);
+        const keysToRemove = [];
+        // 2-step process, so we don't depend on any particular behavior of
+        // key order while removing some
+        for (let i = 0; i < this._storage.length; i++) {
+            const fullKey = this._storage.key(i);
+            if (keyMatch(fullKey)) {
+                keysToRemove.push(fullKey);
+            }
+        }
+        forEach(this._storage.removeItem, keysToRemove);
+    }
+}
+
+class MemStore {
+    constructor() {
+        this._data = {};
+    }
+
+    hasItem(key) {
+        return key in this._data;
+    }
+
+    getItem(key) {
+        // run this storage through JSON too so we know we get a fresh object
+        // each retrieval
+        return JSON.parse(this._data[key] || null);
+    }
+
+    setItem(key, value) {
+        this._data[key] = JSON.stringify(value);
+    }
+
+    removeItem(key) {
+        delete this._data[key];
+    }
+
+    clear(keyPrefix) {
+        forEach(
+            key => delete this._data[key],
+            filter(keyPrefixMatch(keyPrefix), keys(this._data))
+        );
+    }
+}
+
+const stores = {
+    local: new WebStore(window.localStorage),
+    session: new WebStore(window.sessionStorage),
+    memory: new MemStore(),
+};
+
+const noopTransform = {
+    extract: propValue => propValue,
+    apply: (storedValue, _propValue) => storedValue,
+};
+
+const getTransform = (element, propName) =>
+    (element.persistenceTransforms || {})[propName] || noopTransform;
+
+const getNewValKey = (id, propName) => id + '.' + propName;
+const getOriginalValKey = newValKey => newValKey + '.orig';
+const getPersistIdKey = newValKey => newValKey + '.id';
+
+const getProps = layout => {
+    const {props} = layout;
+    const {id, persistence} = props;
+    if (!id || !persistence) {
+        return {props};
+    }
+
+    const element = Registry.resolve(layout);
+    const persisted_props =
+        props.persisted_props || element.defaultProps.persisted_props;
+    const persistence_type =
+        props.persistence_type || element.defaultProps.persistence_type;
+    if (!persisted_props || !persistence_type) {
+        return {};
+    }
+    return {id, props, element, persistence, persisted_props, persistence_type};
+};
+
+export function recordUiEdit(layout, newProps) {
+    const {
+        id,
+        props,
+        element,
+        persistence,
+        persisted_props,
+        persistence_type,
+    } = getProps(layout);
+    if (!persistence) {
```

Comment:
Might want to quick-exit on an undefined `id`.

Reply:
Or maybe that's handled by https://github.com/plotly/dash/pull/903/files#diff-1f9fdd7834ce2c4e31ca7ce6aa890a1fR167 already, but it's not clear.